File: src/test/FileWriteTest.cpp
/**
 * Copyright (c) 2022, Jefferson Science Associates
 *
 * Thomas Jefferson National Accelerator Facility
 * Data Acquisition Group
 *
 * 12000, Jefferson Ave, Newport News, VA 23606
 * Phone : (757)-269-7100
 *
 * @date 07/01/2022
 * @author timmer
 */


#include <cstdint>
#include <memory>
#include <limits>
#include <fstream>

#include "eviocc.h"


namespace evio {


    static void EventWriterTest() {

        //---------------------------------------------
        // Use CompactEventBuilder to create an event
        //---------------------------------------------

        // Create room for entire big event
        size_t bufSize = 1300000;
        CompactEventBuilder ceb(bufSize, ByteOrder::ENDIAN_LOCAL, true);
        ceb.openBank(1, DataType::INT32, 1);
        int32_t dd[250000];
        ceb.addIntData(dd, 250000);
        ceb.closeAll();
        std::shared_ptr<ByteBuffer> bigEvt = ceb.getBuffer();

        // Little event
        bufSize = 1000;
        CompactEventBuilder eb(bufSize, ByteOrder::ENDIAN_LOCAL, true);
        eb.openBank(1, DataType::INT32, 1);
            eb.addIntData(dd, 3);
        eb.closeAll();
        std::shared_ptr<ByteBuffer> littleEvt = eb.getBuffer();

        std::cout << "Buf pos = " << bigEvt->position() << ", lim = " << bigEvt->limit() <<
        ", cap = " << bigEvt->capacity() << std::endl;

        // Write into a file
        uint32_t targetRecordBytes = 900000; // 900KB
        uint32_t bufferBytes = 1000000; // 1MB

        std::string fname = "./codaFileTestCC.ev";

        int evioVersion = 4;

        if (evioVersion == 4) {
            EventWriterV4 writer(fname, "", "", 1, 0, targetRecordBytes,
                                 100000, ByteOrder::ENDIAN_LOCAL, "", true, false,
                                 nullptr, 0, 0, 1, 1, bufferBytes);

            std::cout << "Write little event 1" << std::endl;
            writer.writeEvent(littleEvt, false);
            // Delay between writes
            //std::this_thread::sleep_for(std::chrono::seconds(2));
            std::cout << "Write BIG event 1" << std::endl;
            writer.writeEvent(bigEvt, false);
            //std::this_thread::sleep_for(std::chrono::seconds(2));
            std::cout << "Write little event 2" << std::endl;
            writer.writeEvent(littleEvt, true);
            //std::this_thread::sleep_for(std::chrono::seconds(2));
            std::cout << "Write BIG event 2" << std::endl;
            writer.writeEvent(bigEvt, true);
            std::cout << "WRTER CLOSE" << std::endl;
            writer.close();
        }
        else {
            EventWriter writer(fname, "", "", 1, 0, targetRecordBytes,
                                100000, ByteOrder::ENDIAN_LOCAL, "", true, false,
                                nullptr, 1, 0, 1, 1, Compressor::CompressionType::UNCOMPRESSED,
                                0, 0, bufferBytes);

            std::cout << "Write little event 1" << std::endl;
            writer.writeEventToFile(littleEvt, false, false);
            // Delay between writes
            //std::this_thread::sleep_for(std::chrono::seconds(2));
            std::cout << "Write BIG event 1" << std::endl;
            writer.writeEventToFile(bigEvt, false, false);
            //std::this_thread::sleep_for(std::chrono::seconds(2));
            std::cout << "Write little event 2" << std::endl;
            writer.writeEventToFile(littleEvt, false, true);
            //std::this_thread::sleep_for(std::chrono::seconds(2));
            std::cout << "Write BIG event 2" << std::endl;
            writer.writeEventToFile(bigEvt, false, true);
            std::cout << "WRTER CLOSE" << std::endl;
            writer.close();
        }
    }

}



int main(int argc, char **argv) {
    evio::EventWriterTest();
    return 0;
}

File: java/org/jlab/coda/jevio/test/FileWriteTest.java
package org.jlab.coda.jevio.test;

import org.jlab.coda.hipo.CompressionType;
import org.jlab.coda.jevio.*;

import java.io.IOException;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;

/**
 * Test program. Mirrors the FileWriteTest.cpp file.
 *
 * @author timmer
 * Date: Dec 2, 2024
 */
public class FileWriteTest {



    void eventWriterTest() throws EvioException, IOException {

        //---------------------------------------------
        // Use CompactEventBuilder to create an event
        //---------------------------------------------

        // Create room for entire big event
        int bufSize = 1300000;
        CompactEventBuilder ceb = new CompactEventBuilder(bufSize, ByteOrder.nativeOrder(), true);
        ceb.openBank(1, 1, DataType.INT32);
        int[] dd = new int[250000];
        ceb.addIntData(dd, 0, 250000);
        ceb.closeAll();
        ByteBuffer bigEvt = ceb.getBuffer();

        // Little event
        bufSize = 1000;
        CompactEventBuilder eb = new CompactEventBuilder(bufSize, ByteOrder.nativeOrder(), true);
        eb.openBank(1, 1, DataType.INT32);
        eb.addIntData(dd, 0, 3);
        eb.closeAll();
        ByteBuffer littleEvt = eb.getBuffer();

        System.out.println("Buf pos = " + bigEvt.position() + ", lim = " + bigEvt.limit() + ", cap = " + bigEvt.capacity());

        // Write into a file
        int targetRecordBytes = 900000; // 900KB
        int bufferBytes = 1000000; // 1MB

        String fname = "./codaFileTestCC.ev";

        int evioVersion = 6;

        if (evioVersion == 4) {
            EventWriterV4 writer = new EventWriterV4(fname, null, null, 1, 0, targetRecordBytes,
                    100000, bufferBytes, ByteOrder.nativeOrder(), null, null, true,
                    false, null, 0, 0, 1, 1);

            System.out.println("Write little event 1, buf pos = " + littleEvt.position() + ", lim = " + littleEvt.limit());
            writer.writeEvent(littleEvt, false);
            //writer.writeEventToFile(nullptr, littleEvt, false);
            // Delay between writes
            //std::this_thread::sleep_for(std::chrono::seconds(2));
            System.out.println("Write BIG event 1");
            writer.writeEvent(bigEvt, false);
            //std::this_thread::sleep_for(std::chrono::seconds(2));
            System.out.println("Write little event 2, buf pos = " + littleEvt.position() + ", lim = " + littleEvt.limit());
            writer.writeEvent(littleEvt, true);
            //std::this_thread::sleep_for(std::chrono::seconds(2));
            System.out.println("Write BIG event 2");
            writer.writeEvent(bigEvt, true);
            System.out.println("WRITER CLOSE");
            writer.close();


            EvioReader reader = new EvioReader(fname);
            EvioEvent ev;
            while ( (ev = reader.parseNextEvent()) != null) {

            }
        }
        else {
            EventWriter writer = new EventWriter(fname, null, null, 1, 0, targetRecordBytes,
                    100000, ByteOrder.nativeOrder(), null, true, false,
                    null, 0, 0, 1, 1, CompressionType.RECORD_UNCOMPRESSED,
                    0, 0, bufferBytes);

            System.out.println("Write little event 1, buf pos = " + littleEvt.position() + ", lim = " + littleEvt.limit());
            writer.writeEvent(littleEvt);
            //writer.writeEventToFile(littleEvt, false, false);
            // Delay between writes
            //std::this_thread::sleep_for(std::chrono::seconds(2));
            System.out.println("Write BIG event 1");
            writer.writeEvent(bigEvt);
            //writer.writeEventToFile(bigEvt, false, false);
            //std::this_thread::sleep_for(std::chrono::seconds(2));
            System.out.println("Write little event 2, buf pos = " + littleEvt.position() + ", lim = " + littleEvt.limit());
            writer.writeEvent(littleEvt);
            //writer.writeEventToFile(littleEvt, false, true);
            //std::this_thread::sleep_for(std::chrono::seconds(2));
            System.out.println("Write BIG event 2");
            writer.writeEvent(bigEvt);
            //writer.writeEventToFile(bigEvt, false, true);
            System.out.println("WRITER CLOSE");
            writer.close();
        }
    }



    /**
     * Run as a stand-alone application.
     */
    public static void main(String[] args) {
        FileWriteTest receiver = new FileWriteTest();
        try {
            receiver.eventWriterTest();
        }
        catch (Exception e) {
            e.printStackTrace();
        }
    }



}
